### ISSUE TITLE
[RFC] Optimize Communication and Memory via Resharding, Offloading, and Scheduling

### ISSUE THREAD

**Issue by HzZHoO** [volcengine/verl#2802]
##  **1. Summary**  
This RFC proposes three optimizations to address communication redundancy, memory overhead, and scheduling inefficiencies in veRL:  
1. **Resharding improvements** for expert- (EP) and tensor-parallel (TP) parameters, replacing `AllGather` with `AllToAllV` to reduce communication volume and temporary memory.  
2. **Offload adjustment** to minimize unnecessary host-device transfers during actor updates.  
3. **Device-to-device (D2D) scheduling** with `TensorCache` to eliminate worker startup delays via direct communication between devices.  

**Expected impact**:
- EP reshard: **256×** ↓ in communication/memory (e.g., gen EP256).  
- TP reshard: **2×** ↓ intermediate communication volume (e.g., train PP=8, TP=4).  
- Training tasks: **>93%** ↓ in startup latency on  training engine (e.g., large scale training).  

---

#### **2.1 EP Parameter Resharding**  
**Problem**:  
- Original `AllGather` collects **all 256 experts** per rank during reshard, but only **1 expert** is needed for rollout (EP256) .  
- Causes **14GB temporary memory** per rank (OOM risk) and **255/256 redundant communication** .  

<img width="717" height="351" alt="Image" src="https://github.com/user-attachments/assets/76dcc56a-0674-495c-889d-a37cab38634c" />

**Solution**:  
Replace `AllGather` with `AllToAllV`, ensuring each rank receives **only its assigned expert**.  

<img width="717" height="278" alt="Image" src="https://github.com/user-attachments/assets/653e5e0c-f09f-4244-ae98-87c3a22d2166" />

**Benefits**:  
  - Communication/memory overhead reduced to **1/256** of original.  
  - Eliminates OOM risk.  

**Limitations**:  
- Supports only `moe_tp_extend_ep` scenarios.  
- Requires inference EP divisible by update EP.  

#### **2.2 TP Parameter Resharding**  
**Problem**:  
- Original approach: Broadcast TP slices across PP ranks → `AllGather` in TP group.  
- Inefficiency: Duplicates `AllGather` **PP times** and uses slow broadcast per parameter .  
<img width="717" height="346" alt="Image" src="https://github.com/user-attachments/assets/c3446cde-6a73-4170-8303-6ed1f7decbbd" />

**Solution**:  
1. Reverse reshard order: `AllGather` in **TP group first**, then broadcast in **PP group**.  
2. Replace PP broadcast with **single `AllToAllV`** for all layers.  
<img width="717" height="356" alt="Image" src="https://github.com/user-attachments/assets/0380c893-9d13-4715-ab3c-fa15dd01cda8" />

**Communication Analysis**:
Assume that the model has a total of `n` layers, the total number of TP parameters across all layers is `m`, the number of TP weights in a single layer is `k`, the training PP size is `p`, and the training TP size is `t`.
  | Step | Original | Optimized | Improvement |  
  |---|---|---|---|  
  | Step1 | PP group <br> Comm: `n·m/t`<br>Count: `n·k` | TP group <br> Comm: `n·m/p`<br>Count: `n·k/p` | Comm↓ **50%** (e.g., `p=8,t=4`)<br>TP comm Count↓ `(p-1)/p·n·k` |  
  | Step2 | TP group <br> Comm: `n·m`<br>Count: `n·k` | PP group <br> Comm: `n·m`<br>Count: `n·k/p` | PP comm Count↓ `(p-1)/p·n·k` |  

**Limitations**:  
- Applies only to TP sharded parameters that have counterpart on each pp rank (e.g., excludes dense MLP/embedding weights).  

### **2.3 Future: Unified Resharding**  
**Proposal**: Merge EP&PP and TP&PP resharding into **one communication step** to eliminate intermediate buffers and further reduce latency. *(Design in progress)*  

<img width="717" height="268" alt="Image" src="https://github.com/user-attachments/assets/7a092573-f4d8-4b9a-886e-b8dc065162d5" />

<img width="717" height="212" alt="Image" src="https://github.com/user-attachments/assets/f780e6e7-b4ef-487d-8438-0bb3f50506e8" />

---

##  **3. Offload Optimization**  
**Problem**:  
- actor weights offloaded after `update_actor` → resharing requires reloading, adding **~2x PCIe transfers**.  

**Solution**:  
- **Delay `actor` offload** until after resharding completes.  
- **Benefit**: Saves **one offload and one onload** per training step.  

---

## **4. Model Scheduling Optimization**  
**Problem**:  
- Sequential task dispatch via Ray causes **worker startup skew** (e.g., 96s between worker_0 and worker_255) due to large tensor serialization/deserialization . This results in **>40s communication stalls** of each task on train engine (e.g., `compute_ref_log_prob`).  

<img width="1316" height="283" alt="Image" src="https://github.com/user-attachments/assets/bc3ec377-3f5e-459e-8d8c-359501770dba" />

<img width="470" height="266" alt="Image" src="https://github.com/user-attachments/assets/f684fd88-31d6-49ae-89f9-196f171fa0ce" />

### **4.1 Solution**:  
- **D2D data transfer during scheduling with `TensorCache`**:  
  - Cache intermediate tensors on device.  
  - Replace H2D/D2H transfers with **`AllToAllV` (D2D) for tensor redistribution**.  
  - Minimize host-device transfers to **CPU-critical tensors only**.  

<img width="1697" height="412" alt="Image" src="https://github.com/user-attachments/assets/667509c4-f02e-4d1c-8b33-0adb3a0060aa" />

### **4.2 Optimize data flow**  
<img width="1209" height="787" alt="Image" src="https://github.com/user-attachments/assets/bbf7d176-22d6-4660-b32b-8bd708e4207c" />

### **4.3 TensorCache Design**  
| **Attribute/Method** | **Role** |  
|---|---|  
| `mini_bs` | Batch size per training step → computes source DP size |  
| `tensor_cached` | Dict of cached tensors (key: tensor name) |  
| `cache_tensors()` | Caches tensors from `DataProto` (optional host passthrough) |  
| `get_cached_tensors()` | Returns tensors via `DataProto`; triggers `AllToAllV` if DP ranks mismatch |  
| `clear()` | Flushes cache post-`update_actor` |  
When DP ranks mismatch, the data resharding on batch dim is as follows:

<img width="1331" height="342" alt="Image" src="https://github.com/user-attachments/assets/1ea54193-a261-4461-93c6-d4439ee61640" />

### **4.4 Performance Gains**  
| **Task** | **Original Max Skew** | **Optimized Max Skew** | **Reduction** |  
|---|---|---|---|  
| `compute_ref_log_prob` | ~57s | <1s | 98.2% |  
| `update_actor` | ~73s | ~5s | 93.2% |  

### **4.4 Limitations**:  
- Requires **aligned DP ranks** between tasks.  
- Supports **GRPO on-policy only**.  
- Excludes `balance_batch`, `packed`

**Comment by eric-haibin-lin** [comment]
can u do it with a general implementation without per-model configs?

**Comment by HzZHoO** [comment]
> can u do it with a general implementation without per-model configs?

We are trying to do this, but there may still be some constraints in our implementation

**Comment by OwO-q** [comment]
Do you have a draft implementation or commit that can be shared?

**Comment by huhuiqi7** [comment]
Regarding **Resharding Optimizations**. Is the rollout using vLLM? When vLLM runs with tensor parallelism (TP) enabled, the weight loading functions all take full (unsharded) weight matrices as input. Would this require extensive modifications to vLLM?

**Comment by Kirrito-k423** [comment]
定时清理检查结论：按社区 issue 定时处理规则关闭此 issue。

证据：最近一次非当前运行账号的公开回复或引用是 2025-10-24T07:06:06Z（actor: huhuiqi7, kind: comment），早于本轮 6 个月阈值 2026-01-13T01:10:17.218155+00:00。
检查来源：https://github.com/verl-project/verl/issues/2802，comments API https://api.github.com/repos/verl-project/verl/issues/2802/comments，timeline API https://api.github.com/repos/verl-project/verl/issues/2802/timeline。

如果作者或社区成员仍能提供新的复现、证据或需求背景，请在这里重新评论请求维护者重开，或创建新的 issue 并链接到本 issue。

Scheduled cleanup conclusion: this issue is being closed under the community issue processing policy.

Evidence: The latest public reply or reference from an account other than the current automation account was 2025-10-24T07:06:06Z (actor: huhuiqi7, kind: comment), which is older than the six-month threshold for this run (2026-01-13T01:10:17.218155+00:00).
Sources checked: https://github.com/verl-project/verl/issues/2802, comments API https://api.github.com/repos/verl-project/verl/issues/2802/comments, timeline API https://api.github.com/repos/verl-project/verl/issues/2802/timeline.

If the author or community can still provide a fresh reproduction, evidence, or requirement context, please comment here to ask maintainers to reopen it, or open a new issue and link back to this one.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
